kore/app/share/GlobalMain.hs

```diff
@@ -590,42 +585,51 @@ deserializeDefinition ::
     KoreSolverOptions ->
     FilePath ->
     ModuleName ->
-    UTCTime ->
     Main SerializedDefinition
 deserializeDefinition
     simplifierx
     solverOptions
     definitionFilePath
-    mainModuleName
-    exeLastModifiedTime =
+    mainModuleName =
         do
-            bytes <- ByteString.readFile definitionFilePath & liftIO
-            let magicBytes = ByteString.drop 8 $ ByteString.take 16 bytes
-            let magic = Binary.decode @Word64 magicBytes
-            case magic of
-                -- This magic number comes from the Data.Compact.Serialize moduile source:
-                -- https://hackage.haskell.org/package/compact-0.2.0.0/docs/src/Data.Compact.Serialize.html#magicNumber
-                -- The field is not exported by the package so we have to manually specify it here.
-                -- If you update the version of the compact package, you should double check that
-                -- the file format has not changed. They don't provide any particular guarantees
-                -- of stability across versions because the serialized data becomes invalid when
-                -- any changes are made to the binary at all, so there would be no point.
-                --
-                -- We use this magic number to detect if the input file for the definition is
-                -- a serialized Data.Compact region or if it is a textual KORE definition.
-                0x7c155e7a53f094f2 -> do
-                    defnLastModifiedTime <- getModificationTime definitionFilePath & liftIO
-                    if defnLastModifiedTime < exeLastModifiedTime
-                        then errorOutOfDate "serialized definition is out of date. Rerun kompile or kore-exec --serialize."
-                        else do
-                            result <- unsafeReadCompact definitionFilePath & liftIO
-                            either errorParse (return . getCompact) result
-                _ ->
+            maybeSerializedDefinition <-
+                withFile definitionFilePath ReadMode readContents & liftIO
+            case maybeSerializedDefinition of
+                Just serializedDefinition ->
+                    return serializedDefinition
+                Nothing ->
                     makeSerializedDefinition
                         simplifierx
                         solverOptions
                         definitionFilePath
                         mainModuleName
+      where
+        readContents definitionHandle = do
+            fingerprint <-
+                ByteString.hGet definitionHandle 32
+                    <&> Binary.decode
+            magicNumber <-
+                ByteString.hGet definitionHandle 16
+                    <&> ByteString.drop 8
+                    <&> Binary.decode @Word64
+            case magicNumber of
+                0x7c155e7a53f094f2 -> do
+                    checkFingerprint fingerprint
+                    hSeek definitionHandle AbsoluteSeek 32
```

Comment:
I think you would not need to check the magic number by yourself here, `hUnsafeGetCompact` will do that (by way of `Binary.decodeOrFail` calling [`Binary.get`](https://hackage.haskell.org/package/compact-0.2.0.0/docs/src/Data.Compact.Serialize.html#local-8214565720323790560). 
That would save rewinding the handle.

Reply:
I think the reason this check was added was because depending on whether that magic number exists we know if the file is binary or textual, so we can delegate to the appropriate procedure.